.circleci/config.yml

```
@@ -2085,16 +2064,112 @@ jobs:
       - run:
           name: Build kona host
           command: just build-kona-host
-          working_directory: kona
+          working_directory: kona-proofs
       - save_cache:
-          key: kona-host-{{ checksum "./kona/justfile" }}-{{ checksum "./kona/version.json" }}
+          key: kona-host-{{ checksum "./kona-proofs/justfile" }}-{{ checksum "./kona-proofs/version.json" }}
           name: Save Kona host to cache
           paths:
-            - "kona/bin/kona-host"
+            - "kona-proofs/bin/kona-host"
       - persist_to_workspace:
           root: .
           paths:
-            - "kona/bin/kona-host"
+            - "kona-proofs/bin/kona-host"
+
+  # Build a single Rust binary for sysgo tests (parameterized).
+  rust-binary-build:
+    docker:
+      - image: <<pipeline.parameters.default_docker_image>>
+    resource_class: xlarge
+    parameters:
+      cache_name:
+        description: "Identifier for cache key and logging"
+        type: string
+      directory:
+        description: "Directory containing the Cargo workspace"
+        type: string
+      build_command:
+        description: "Cargo build command to run"
+        type: string
+      check_binary:
+        description: "Binary name to check for cache hit (in target/release/)"
+        type: string
+      needs_clang:
+        description: "Whether to install clang (needed by bindgen for reth-mdbx-sys)"
+        type: boolean
+        default: false
+    steps:
+      - utils/checkout-with-mise:
+          checkout-method: blobless
+      - run:
+          name: Get << parameters.directory >> submodule SHA
+          command: |
+            SUBMOD_SHA=$(git ls-tree HEAD << parameters.directory >> | awk '{print $3}')
+            echo "export SUBMOD_SHA=$SUBMOD_SHA" >> "$BASH_ENV"
+            echo "Submodule SHA: $SUBMOD_SHA"
+      - restore_cache:
+          name: Restore << parameters.cache_name >> cache
+          keys:
+            - rust-<< parameters.cache_name >>-v3-{{ .Environment.SUBMOD_SHA }}
```

```suggestion
            echo -n "Submodule SHA: "
            git ls-tree HEAD << parameters.directory >> | awk '{print $3}' | tee .submod_sha
      - restore_cache:
          name: Restore << parameters.cache_name >> cache
          keys:
            - rust-<< parameters.cache_name >>-v3-{{ checksum ".submod_sha" }}
```
may fix broken cache key. Also needs to be adapted below in `save_cache` step.